### DISCUSSION TITLE
Question about Signer CRDT

### DISCUSSION THREAD

**Discussion by Ankarrr** [farcasterxyz/protocol#179] [Q&A] [answered]
Hey, I'm reading the specs and I saw `Signer CRDT` has been mentioned multiple times in the CRDTs section but I didn't see the definition of `Signer CRDT`. I'm wondering whether `Signer CRDT` is still valid or it has been moved on-chain. Thanks!

https://github.com/farcasterxyz/protocol/blob/main/docs/SPECIFICATION.md#312-userdata-crdt

**Comment by horsefacts** [comment, marked as answer]
Signers moved on chain to the `KeyRegistry` contract as part of [FIP-7](https://github.com/farcasterxyz/protocol/discussions/103). You're right, there are still some references to the old CRDT in the spec doc.

The new equivalent to "`m.signer` must be a valid Signer in the add-set of the Signer CRDT for `message.fid`" is "`m.signer` must be a valid key with `Keystate.ADDED` in the `KeyRegistry` contract for `message.fid`."

**Comment by Ankarrr** [comment]
I see. Thanks for your explaination!

**Comment by Ankarrr** [comment]
I created a PR https://github.com/farcasterxyz/protocol/pull/186 to update the spec. I hope this is helpful!
